fairscale/nn/moe/moelayer.py

```diff
@@ -35,21 +36,25 @@ class MOELayer(Base):
             expert network
     """
 
-    def __init__(self, gate: Module, expert: Module) -> None:
+    def __init__(self, gate: Module, expert: Module, group: Optional[Any] = None) -> None:
         super().__init__()
         self.gate = gate
         self.expert = expert
+        self.group = group if group is not None else dist.group.WORLD
+        self.world_size = dist.get_world_size(self.group)
 
     def all_to_all_dispatch(self, dispatch_mask: Tensor, input: Tensor) -> Tensor:
         dispatched_input = torch.einsum("gsec,gsm->egcm", dispatch_mask.float(), input)
-        # TODO(msb) all-to-all
-        dispatched_input = torch.squeeze(dispatched_input, 0)  # drop E dimension
+        dispatched_input = dispatched_input.contiguous()
+        chunks = list(dispatched_input.chunk(self.world_size))
+        dist.all_to_all(chunks, chunks, self.group)
         return dispatched_input
 
     def all_to_all_combine(self, combine_weights: Tensor, input: Tensor) -> Tensor:
-        # TODO(msb) all-to-all
-        expert_output = torch.unsqueeze(input, 1)  # add E dimension
-        output = torch.einsum("gsec,gecm->gsm", combine_weights, expert_output)
+        expert_output = input.contiguous()
```

Comment:
Similarly, here, since we are using all_to_all() for the first time, it would be nice to have a basic test to ensure expected tensors are combined at the expected rank.

Reply:
test_moelayer.py::test_forward does this

The `assert torch.allclose(input, output)` verifies that we re-assembled the input after passing it through an identity matrix. I had to use allclose because the input tensor is split across experts.